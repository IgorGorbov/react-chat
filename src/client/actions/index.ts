import {
  CLIENT_CONNECT,
  REGISTER_CLIENT,
  FETCH_ALL_USERS,
  FETCH_USER_CHATS,
  SEARCH_CHAT,
  CHANGE_USER_STATUS,
  CHANGE_CHAT,
  ADD_CHAT,
  NEW_CHAT,
  DELETE_CHAT,
  ADD_MESSAGE,
  NEW_MESSAGE,
  CHOOSE_USER,
  SELECT_MESSAGE,
  DELETE_MESSAGES,
  READ_MESSAGES,
  REFRESH_MESSAGES,
} from '../constants/actionTypes';

export const clientConnect = () => ({
  type: CLIENT_CONNECT,
});

export const changeUserStatus = (status: string) => ({
  type: CHANGE_USER_STATUS,
  payload: status,
});

export const changeChat = (chat: { id: number; companion: IUser }) => ({
  type: CHANGE_CHAT,
  payload: {
    id: chat.id,
    companion: chat.companion,
  },
});

export const fetchAllUsers = (users: IUser[]) => ({
  type: FETCH_ALL_USERS,
  payload: users,
});

export const fetchUserChats = (chats: IChat[]) => ({
  type: FETCH_USER_CHATS,
  payload: chats,
});

export const search = (text: string) => ({
  type: SEARCH_CHAT,
  payload: text,
});

export const addChat = (companion: IUser) => ({
  type: ADD_CHAT,
  payload: companion,
});

export const newChat = (currentUser: IUser, companion: IUser) => ({
  type: NEW_CHAT,
  payload: { currentUser, companion },
});

export const deleteChat = (chatId: number) => ({
  type: DELETE_CHAT,
  payload: chatId,
});

export const addMessage = (text: string) => ({
  type: ADD_MESSAGE,
  payload: text,
});

export const newMessage = (chatId: number, message: IMessage) => ({
  type: NEW_MESSAGE,
  payload: {
    chatId,
    message,
  },
});

export const registerClient = (name: string) => ({
  type: REGISTER_CLIENT,
  payload: name,
});

export const chooseUser = (user: IUser) => ({
  type: CHOOSE_USER,
  payload: user,
});

export const selectMessage = (messageId: number) => ({
  type: SELECT_MESSAGE,
  payload: messageId,
});

export const deleteMessages = (
  selectedMessages: number[],
  currentUserID: number,
  currentChatId: number,
) => ({
  type: DELETE_MESSAGES,
  payload: { selectedMessages, currentUserID, currentChatId },
});

export const refreshMessages = (messages: number[], currentChatId: number) => ({
  type: REFRESH_MESSAGES,
  payload: { messages, currentChatId },
});

export const readMessages = (currentChatId: number, currentUserID: number) => ({
  type: READ_MESSAGES,
  payload: { currentUserID, currentChatId },
});
