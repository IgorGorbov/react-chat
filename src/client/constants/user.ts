export enum Status {
  Online = 'Online',
  Offline = 'Offline',
  Away = 'Away',
  Busy = 'Busy',
}

export const statusColors = {
  [Status.Online]: 'green',
  [Status.Offline]: 'grey',
  [Status.Away]: '#dede0f',
  [Status.Busy]: 'red',
};
