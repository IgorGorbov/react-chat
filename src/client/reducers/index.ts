import { combineReducers } from 'redux';
import session from './session';
import chats from './chats';
import filters from './filters';
import currentChat from './currentChat';
import users from './users';

export default combineReducers({
  session,
  users,
  chats,
  currentChat,
  filters,
});
