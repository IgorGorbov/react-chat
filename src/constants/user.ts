export const defaultAvatar = 'https://emilcarlsson.se/assets/mikeross.png';
export enum Status {
  Online = "Online",
  Offline = "Offline",
  Away = "Away",
  Busy = "Busy"
}

export const statusColors = {
  [Status.Online]: 'green',
  [Status.Offline]: 'grey',
  [Status.Away]: 'yellow',
  [Status.Busy]: 'red'
};
