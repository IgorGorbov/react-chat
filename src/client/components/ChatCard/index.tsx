import * as React from 'react';
import {
  UserCardContainer,
  ImageContainer,
  UserAvatar,
  UserStatus,
  UserInfoContainer,
  UserName,
  TextMessage,
  NewCountMessage,
} from '../../styledComponents/index';
import { statusColors } from '../../constants/user';

interface IChatCardProps {
  user: IUser;
  key: number;
  chat: IChat;
  isActive: string;
  onChangeChat: () => void;
}

const ChatCard = ({ user, chat, onChangeChat, isActive }: IChatCardProps) => {
  const inBox: any = chat.messages.filter(
    (message: IMessage) => message.to.id !== user.id,
  );
  const lastMessage = inBox.length ? [...inBox].pop().text.slice(0, 10) : '';
  const countNewMessages = inBox.reduce(
    (acc: number, message: IMessage) => acc + Number(!message.isRead),
    0,
  );
  return (
    <UserCardContainer className={isActive} onClick={onChangeChat}>
      <ImageContainer>
        <UserAvatar
          maxwidth="65px"
          maxheight="65px"
          bordersize="1px"
          bordercolor="white"
          src={user.avatar}
          size="tiny"
          circular={true}
        />
        <UserStatus color={statusColors[user.status]} />
      </ImageContainer>
      <UserInfoContainer>
        <UserName>{user.name}</UserName>
        <TextMessage>{lastMessage}</TextMessage>
      </UserInfoContainer>
      <NewCountMessage>
        {countNewMessages !== 0 ? countNewMessages : ''}
      </NewCountMessage>
    </UserCardContainer>
  );
};

export default ChatCard;
