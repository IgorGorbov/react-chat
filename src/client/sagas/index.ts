import { SagaIterator } from 'redux-saga';
import { all, takeEvery, call, select } from 'redux-saga/effects';
import * as types from '../constants/actionTypes';
import {
  getCompanion,
  getChatCurrentId,
  getCurrentUser,
} from '../selectors/index';

const clientConnect = function*(socket: ISocket): SagaIterator {
  yield call(socket.getUsers, 'available', (err: any, users: IUser[]) => {
    if (err) {
      console.error('clientConnect saga => getUsers', err);
      return;
    }
    socket.dispatchAction({ type: types.FETCH_ALL_USERS, payload: users });
  });
};

const registerClient = function*(
  socket: ISocket,
  action: { payload: string },
): SagaIterator {
  yield call(socket.register, action.payload, (err: any, user: IUser) => {
    if (err) {
      console.error('registerClient saga => register', err);
      return;
    }
    socket.dispatchAction({ type: types.CHOOSE_USER, payload: user });
  });

  yield call(socket.getUserChats, (err: any, chats: IChat[]) => {
    if (err) {
      console.error('registerClient saga => getUserChats', err);
      return;
    }
    socket.dispatchAction({ type: types.FETCH_USER_CHATS, payload: chats });
  });

  yield call(socket.getUsers, 'all', (err: any, users: IUser[]) => {
    if (err) {
      console.error('registerClient saga => getUsers', err);
      return;
    }
    socket.dispatchAction({ type: types.FETCH_ALL_USERS, payload: users });
  });
};

const addChat = function*(
  socket: ISocket,
  action: { payload: IUser },
): SagaIterator {
  yield call(socket.addChat, action.payload, (err: any, chat: IChat) => {
    if (err) {
      console.error('addChat saga => addChat', err);
      return;
    }
    socket.dispatchAction({ type: types.NEW_CHAT, payload: chat });
  });
};

const deleteChat = function*(
  socket: ISocket,
  action: { payload: number },
): SagaIterator {
  yield call(socket.deleteChat, action.payload);
};

const addMessage = function*(
  socket: ISocket,
  action: { payload: string },
): SagaIterator {
  const text = action.payload;
  const companion = yield select(getCompanion);
  const chatId = yield select(getChatCurrentId);

  yield call(socket.addMessage, text, chatId, companion);
};

const deleteMessages = function*(
  socket: ISocket,
  action: { payload: number[] },
): SagaIterator {
  const currentChatId = yield select(getChatCurrentId);

  yield call(
    socket.deleteMessages,
    action.payload,
    (err: any, messages: IMessage[]) => {
      if (err) {
        console.error('deleteMessages saga => deleteMessages', err);
        return;
      }
      socket.dispatchAction({
        type: types.REFRESH_MESSAGES,
        payload: { currentChatId, messages },
      });
    },
  );
};

const changeUserStatus = function*(socket: ISocket): SagaIterator {
  const currentUser = yield select(getCurrentUser);
  yield call(socket.changeUserStatus, currentUser);
};

export default function* SocketSaga(socket: ISocket): SagaIterator {
  yield all([
    takeEvery(types.CLIENT_CONNECT, clientConnect, socket),
    takeEvery(types.REGISTER_CLIENT, registerClient, socket),
    takeEvery(types.CHANGE_USER_STATUS, changeUserStatus, socket),
    takeEvery(types.ADD_CHAT, addChat, socket),
    takeEvery(types.DELETE_CHAT, deleteChat, socket),
    takeEvery(types.ADD_MESSAGE, addMessage, socket),
    takeEvery(types.DELETE_MESSAGES, deleteMessages, socket),
  ]);
}
