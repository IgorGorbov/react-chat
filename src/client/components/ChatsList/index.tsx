import * as React from 'react';
import ChatCard from '../ChatCard/index';
import { UserListWrapper } from '../../styledComponents/index';

function isChatArray(arr: any): arr is [] {
  return arr.length > 0 && typeof arr[0] === 'object' && Array.isArray(arr);
}

interface IUserListProps {
  chats: IChat[];
  currentChatId: number;
  currentUserId: number;
  changeChat: (
    chat: object,
  ) => {
    type: string;
    payload: {
      id: number;
      companion: IUser;
    };
  };
  readMessages: (
    currentChatId: number,
    currentUserID: number,
  ) => {
    type: string;
    payload: { currentChatId: number; currentUserID: number };
  };
}

class ChatsList extends React.Component<IUserListProps> {
  public onChangeChat(chat: IChat, user: IUser) {
    this.props.changeChat({ id: chat.id, companion: user });
    this.props.readMessages(chat.id, this.props.currentUserId);
  }

  public renderChatCard(chat: IChat): JSX.Element | null {
    const { currentUserId, currentChatId } = this.props;
    const user: IUser | undefined = chat.users.find(
      (u: IUser) => u.id !== currentUserId,
    );
    if (!user) {
      return null;
    }
    const isActive = currentChatId === chat.id ? 'active' : '';
    return (
      <ChatCard
        key={user.id}
        user={user}
        chat={chat}
        onChangeChat={() => this.onChangeChat(chat, user)}
        isActive={isActive}
      />
    );
  }

  public render() {
    const chats = Object.values(this.props.chats);
    return (
      <UserListWrapper>
        {isChatArray(chats)
          ? chats.map(chat => this.renderChatCard(chat))
          : null}
      </UserListWrapper>
    );
  }
}

export default ChatsList;
