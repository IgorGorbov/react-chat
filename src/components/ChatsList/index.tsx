import * as React from 'react';
import ChatCard from '../ChatCard';
import { UserListWrapper } from '../../styledComponents';

function isChatArray(arr: any): arr is [] {
  return arr.length > 0 && typeof arr[0] === 'object' && Array.isArray(arr);
}

interface IUserListProps {
  chats: IChat[];
  currentChatId: number;
  currentUserId: number;
  changeChat: (
    chat: object,
  ) => {
    type: string;
    payload: {
      id: number;
      companion: IUser;
    };
  };
}

class ChatsList extends React.Component<IUserListProps> {
  public onChangeChat(chat: IChat, user: IUser) {
    this.props.changeChat({ id: chat.id, companion: user });
  }

  public renderChatCard(chat: IChat): JSX.Element | null {
    const { currentUserId, currentChatId } = this.props;
    const user: IUser | undefined = chat.users.find(
      (u: IUser) => u.id !== currentUserId,
    );
    if (!user) {
      return null;
    }
    const isActive = currentChatId === chat.id ? 'active' : '';
    return (
      <ChatCard
        key={user.id}
        user={user}
        chat={chat}
        onChangeChat={() => this.onChangeChat(chat, user)}
        isActive={isActive}
      />
    );
  }

  public render() {
    const chats = Object.values(this.props.chats);
    return (
      <UserListWrapper>
        {isChatArray(chats)
          ? chats.map(chat => this.renderChatCard(chat))
          : null}
      </UserListWrapper>
    );
  }
}

export default ChatsList;
