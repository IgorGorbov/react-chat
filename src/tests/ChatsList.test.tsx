import * as React from 'react';
import toJSON from 'enzyme-to-json';
import { mount } from 'enzyme';

import ChatsList from '../client/components/ChatsList';
import ChatCard from '../client/components/ChatCard';
import { Status } from '../client/constants/user';
import { UserListWrapper } from '../client/styledComponents';

describe('testing ChatsList', () => {
  interface IUserListProps {
    chats: IChat[];
    currentChatId: number;
    currentUserId: number;
    changeChat: (
      chat: object,
    ) => {
      type: string;
      payload: {
        id: number;
        companion: IUser;
      };
    };
    readMessages: (
      currentChatId: number,
      currentUserID: number,
    ) => {
      type: string;
      payload: { currentChatId: number; currentUserID: number };
    };
  }
  const FakeChangeChat = jest.fn();
  const FakeReadMessages = jest.fn();

  const initState: IUserListProps = {
    chats: [
      {
        id: 1,
        users: [
          {
            id: 1,
            name: 'test',
            avatar: 'test',
            status: Status.Online,
          },
          {
            id: 2,
            name: 'test2',
            avatar: 'test2',
            status: Status.Online,
          },
        ],
        messages: [],
      },
    ],
    currentChatId: 1,
    currentUserId: 1,
    changeChat: FakeChangeChat,
    readMessages: FakeReadMessages,
  };

  let wrapper: any;

  beforeAll(() => {
    wrapper = mount(<ChatsList {...initState} />);
  });

  it('should contain UserListWrapper', () => {
    expect(wrapper.find(UserListWrapper)).toHaveLength(1);
  });

  it('should contain ChatCard', () => {
    const countChat = initState.chats.length;
    const chatCard = wrapper.find(ChatCard);

    expect(chatCard).toHaveLength(countChat);

    chatCard.simulate('click');
    expect(FakeChangeChat).toHaveBeenCalled();
    expect(FakeReadMessages).toHaveBeenCalled();
  });

  it('snapshot ChatsList', () => {
    const tree = toJSON(wrapper);
    expect(tree).toMatchSnapshot();
  });
});
