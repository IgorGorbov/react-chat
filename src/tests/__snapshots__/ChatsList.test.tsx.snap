// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ChatsList snapshot ChatsList 1`] = `
<ChatsList
  changeChat={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "companion": Object {
              "avatar": "test2",
              "id": 2,
              "name": "test2",
              "status": "Online",
            },
            "id": 1,
          },
        ],
      ],
    }
  }
  chats={
    Array [
      Object {
        "id": 1,
        "messages": Array [],
        "users": Array [
          Object {
            "avatar": "test",
            "id": 1,
            "name": "test",
            "status": "Online",
          },
          Object {
            "avatar": "test2",
            "id": 2,
            "name": "test2",
            "status": "Online",
          },
        ],
      },
    ]
  }
  currentChatId={1}
  currentUserId={1}
  readMessages={
    [MockFunction] {
      "calls": Array [
        Array [
          1,
          1,
        ],
      ],
    }
  }
>
  <styled.div>
    <div
      className="sc-cSHVUG cliaFh"
    >
      <ChatCard
        chat={
          Object {
            "id": 1,
            "messages": Array [],
            "users": Array [
              Object {
                "avatar": "test",
                "id": 1,
                "name": "test",
                "status": "Online",
              },
              Object {
                "avatar": "test2",
                "id": 2,
                "name": "test2",
                "status": "Online",
              },
            ],
          }
        }
        isActive="active"
        key="2"
        onChangeChat={[Function]}
        user={
          Object {
            "avatar": "test2",
            "id": 2,
            "name": "test2",
            "status": "Online",
          }
        }
      >
        <styled.div
          className="active"
          onClick={[Function]}
        >
          <div
            className="active sc-kAzzGY dxEwPq"
            onClick={[Function]}
          >
            <styled.div>
              <div
                className="sc-chPdSV gWxTLm"
              >
                <Styled(Image)
                  bordercolor="white"
                  bordersize="1px"
                  circular={true}
                  maxheight="65px"
                  maxwidth="65px"
                  size="tiny"
                  src="test2"
                >
                  <Image
                    as="img"
                    bordercolor="white"
                    bordersize="1px"
                    circular={true}
                    className="sc-ifAKCX irlmET"
                    maxheight="65px"
                    maxwidth="65px"
                    size="tiny"
                    src="test2"
                    ui={true}
                  >
                    <img
                      bordercolor="white"
                      bordersize="1px"
                      className="ui tiny circular image sc-ifAKCX irlmET"
                      maxheight="65px"
                      maxwidth="65px"
                      src="test2"
                    />
                  </Image>
                </Styled(Image)>
                <styled.span
                  color="green"
                >
                  <span
                    className="sc-kgoBCf gMvtpI"
                    color="green"
                  />
                </styled.span>
              </div>
            </styled.div>
            <styled.div>
              <div
                className="sc-kGXeez dCpFuT"
              >
                <styled.p>
                  <p
                    className="sc-kpOJdX bxSrEP"
                  >
                    test2
                  </p>
                </styled.p>
                <styled.p>
                  <p
                    className="sc-fjdhpX gAKYZH"
                  />
                </styled.p>
              </div>
            </styled.div>
            <styled.p>
              <p
                className="sc-dxgOiQ eGtYgX"
              />
            </styled.p>
          </div>
        </styled.div>
      </ChatCard>
    </div>
  </styled.div>
</ChatsList>
`;
